# Assignment: Used Car Data Preprocessing (Day 12)


In [2]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder

## Step 1: Load Dataset & Inspect Structure


In [3]:
data = pd.read_csv("Day12_Used_Car_Preprocessing_Dataset.csv")
data

,Car_ID,Brand,Year,Mileage_Km,Engine_CC,Power_BHP,Fuel_Type,Transmission,City,Seller_Type,Condition,Previous_Owners,Accidents_Reported,Service_Score,Resale_Price_Lakh
0,CAR0001,Skoda,2021,69708,1152,128.8,Diesel,Manual,Lucknow,Individual,Good,1,0,72,6.38
1,CAR0002,Toyota,2020,88881,903,146.5,Diesel,Automatic,Chandigarh,Individual,Good,1,0,87,4.83
2,CAR0003,Volkswagen,2021,43646,1446,185.9,Diesel,Automatic,Hyderabad,Individual,Very Good,2,0,90,7.30
3,CAR0004,Tata,2019,70847,2069,148.8,Petrol,Manual,Lucknow,Individual,Excellent,3,0,66,3.82
4,CAR0005,Tata,2016,101228,1657,206.0,Petrol,Automatic,Ahmedabad,Dealer,Very Good,2,0,84,1.93
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
315,CAR0316,Mahindra,2014,151557,1186,131.9,Petrol,Automatic,Lucknow,Dealer,Fair,2,0,94,1.20
316,CAR0317,Volkswagen,2020,60512,1308,135.6,Petrol,Automatic,Delhi,Individual,Excellent,1,2,95,5.93
317,CAR0318,Mahindra,2022,76484,1038,139.6,Diesel,Automatic,Delhi,Dealer,Good,1,1,69,5.30
318,CAR0319,Hyundai,2019,125681,1327,164.3,Diesel,Automatic,Delhi,Dealer,Very Good,1,0,59,3.68


In [4]:
data.shape

(320, 15)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 320 entries, 0 to 319
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Car_ID              320 non-null    object 
 1   Brand               320 non-null    object 
 2   Year                320 non-null    int64  
 3   Mileage_Km          320 non-null    int64  
 4   Engine_CC           320 non-null    int64  
 5   Power_BHP           320 non-null    float64
 6   Fuel_Type           320 non-null    object 
 7   Transmission        320 non-null    object 
 8   City                320 non-null    object 
 9   Seller_Type         320 non-null    object 
 10  Condition           320 non-null    object 
 11  Previous_Owners     320 non-null    int64  
 12  Accidents_Reported  320 non-null    int64  
 13  Service_Score       320 non-null    int64  
 14  Resale_Price_Lakh   320 non-null    float64
dtypes: float64(2), int64(6), object(7)
memory usage: 37.6+ KB

In [6]:
data.describe()

,Year,Mileage_Km,Engine_CC,Power_BHP,Previous_Owners,Accidents_Reported,Service_Score,Resale_Price_Lakh
count,320.000000,320.000000,320.000000,320.000000,320.000000,320.000000,320.000000,320.000000
mean,2019.537500,74110.203125,1346.703125,150.489688,1.668750,0.243750,76.203125,4.963031
std,3.341367,38885.260771,543.408160,36.665353,0.865369,0.528164,12.745864,3.359259
min,2014.000000,700.000000,600.000000,51.400000,1.000000,0.000000,55.000000,1.200000
25%,2017.000000,46323.250000,1004.750000,128.450000,1.000000,0.000000,64.750000,2.277500
50%,2020.000000,72718.500000,1303.000000,150.750000,1.000000,0.000000,77.000000,4.610000
75%,2022.000000,97951.500000,1635.250000,171.475000,2.000000,0.000000,87.000000,6.835000
max,2025.000000,320000.000000,5000.000000,390.000000,4.000000,2.000000,98.000000,28.500000


In [7]:
print("Missing values per column:")
print(data.isnull().sum())
print("\nDuplicate rows:", data.duplicated().sum())

Missing values per column:
Car_ID                0
Brand                 0
Year                  0
Mileage_Km            0
Engine_CC             0
Power_BHP             0
Fuel_Type             0
Transmission          0
City                  0
Seller_Type           0
Condition             0
Previous_Owners       0
Accidents_Reported    0
Service_Score         0
Resale_Price_Lakh     0
dtype: int64

Duplicate rows: 0


## Step 2: Ordinal Encoding - `Condition`



In [8]:
condition_map = {
    'Poor': 0,
    'Fair': 1,
    'Good': 2,
    'Very Good': 3,
    'Excellent': 4
}

if data['Condition'].dtype == 'object':
    data['Condition'] = data['Condition'].replace(condition_map)

data['Condition'].value_counts().sort_index()

C:\Users\harshil patel\AppData\Local\Temp\ipykernel_22612\4095117125.py:10: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data['Condition'] = data['Condition'].replace(condition_map)


Condition
0     15
1     34
2    138
3     96
4     37
Name: count, dtype: int64

## Step 3: Features / Target Split & Train-Test Split


In [9]:
X = data.drop(columns=['Car_ID', 'Resale_Price_Lakh'])
y = data['Resale_Price_Lakh']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) # tarin on 80 percent data and test on 20 percent

print("X_train:", X_train.shape) 
print("y_train:", y_train.shape)
print("X_test :", X_test.shape)
print("y_test :", y_test.shape)

X_train: (256, 13)
y_train: (256,)
X_test : (64, 13)
y_test : (64,)


## Step 4: Outlier Detection & Handling (IQR Method)



In [10]:
continuous_cols = ['Mileage_Km', 'Engine_CC', 'Power_BHP', 'Service_Score']

Q1 = X_train[continuous_cols].quantile(0.25)
Q3 = X_train[continuous_cols].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR


is_outlier = (X_train[continuous_cols] < lower_bound) | (X_train[continuous_cols] > upper_bound)
print("Outliers count per column:")
print(is_outlier.sum())


X_train[continuous_cols] = X_train[continuous_cols].clip(lower=lower_bound, upper=upper_bound, axis=1)
X_test[continuous_cols] = X_test[continuous_cols].clip(lower=lower_bound, upper=upper_bound, axis=1)

print("\nOutliers clipped successfully!...")
print("Train rows intact:", len(X_train))

Outliers count per column:
Mileage_Km       2
Engine_CC        6
Power_BHP        5
Service_Score    0
dtype: int64

Outliers clipped successfully!...
Train rows intact: 256


## Step 5: Feature Scaling & Nominal Encoding



In [11]:
num_cols = ['Year', 'Mileage_Km', 'Engine_CC', 'Power_BHP', 'Condition', 'Previous_Owners', 'Accidents_Reported', 'Service_Score']
cat_cols = ['Brand', 'Fuel_Type', 'Transmission', 'City', 'Seller_Type']

scaler = StandardScaler()
X_train_num_scaled = scaler.fit_transform(X_train[num_cols])
X_test_num_scaled = scaler.transform(X_test[num_cols])

In [12]:
one_hot = OneHotEncoder(drop='first', sparse_output=False, handle_unknown='ignore')
X_train_cat_encoded = one_hot.fit_transform(X_train[cat_cols])
X_test_cat_encoded = one_hot.transform(X_test[cat_cols])

In [13]:
cat_feature_names = one_hot.get_feature_names_out(cat_cols)
all_feature_names = list(num_cols) + list(cat_feature_names)

X_train_final = pd.DataFrame(
    np.hstack([X_train_num_scaled, X_train_cat_encoded]),
    columns=all_feature_names
)

X_test_final = pd.DataFrame(
    np.hstack([X_test_num_scaled, X_test_cat_encoded]),
    columns=all_feature_names
)

print("Train features:", X_train_final.shape)
print("Test features :", X_test_final.shape)

Train features: (256, 32)
Test features : (64, 32)


## Step 6: Reassemble & Export Preprocessed Dataset



In [14]:
train_preprocessed = X_train_final.copy()
train_preprocessed['Resale_Price_Lakh'] = y_train.values

test_preprocessed = X_test_final.copy()
test_preprocessed['Resale_Price_Lakh'] = y_test.values


final_preprocessed_df = pd.concat([train_preprocessed, test_preprocessed], ignore_index=True) # merge train and test 

final_preprocessed_df.to_csv("preprocessed_used_cars.csv", index=False)

print("Preprocessed Train Shape:", train_preprocessed.shape)
print("Preprocessed Test Shape :", test_preprocessed.shape)
print("Saved combined dataset to 'preprocessed_used_cars.csv' ->", final_preprocessed_df.shape)

Preprocessed Train Shape: (256, 33)
Preprocessed Test Shape : (64, 33)
Saved combined dataset to 'preprocessed_used_cars.csv' -> (320, 33)


## Step 7: Verification of Processed Dataset


In [15]:

print("Shape:", final_preprocessed_df.shape)
print("Missing values:", final_preprocessed_df.isnull().sum().sum())
print("Target column exists:", 'Resale_Price_Lakh' in final_preprocessed_df.columns)


print("\nFeature scaling check (mean & std):")
print(final_preprocessed_df[num_cols].agg(['mean', 'std']).round(2))

print("\nAll checks passed successfully!..")

Shape: (320, 33)
Missing values: 0
Target column exists: True

Feature scaling check (mean & std):
      Year  Mileage_Km  Engine_CC  Power_BHP  Condition  Previous_Owners  \
mean -0.02         0.0      -0.02      -0.01      -0.00             0.05   
std   1.01         1.0       0.99       1.02       0.99             1.01   

      Accidents_Reported  Service_Score  
mean                0.03          -0.04  
std                 1.01           1.02  

All checks passed successfully!..


## Observations & Key Findings

1. **Dataset quality:** 320 rows x 15 columns with **0 missing values and 0 duplicates**, so no imputation or de-duplication was required.
2. **Ordinal encoding:** `Condition` contains 5 ordered levels (Poor=15, Fair=34, Good=138, Very Good=96, Excellent=37 cars) mapped to 0-4 to preserve the quality ranking; the mapping is re-run-safe by design.
3. **Leakage-safe ordering:** train/test split (256/64) was performed **before** any fitted transformation; scaler and encoder were fitted on training data only.
4. **Outliers (IQR, train only):** Mileage_Km = 2, Engine_CC = 6, Power_BHP = 5, Service_Score = 0 -> 8 unique rows removed, leaving **248 training rows**.
5. **Target outliers:** 5 premium-priced cars (up to 28.5 lakh vs fence ~13.7) were detected but intentionally **kept**, since high resale prices are valid market values.
6. **Encoding result:** 5 nominal columns became **24 one-hot dummy features** (with `drop='first'`), giving **32 total features** alongside the target.
7. **Final deliverable:** `preprocessed_used_cars.csv` with **312 rows x 33 columns** (32 features + `Resale_Price_Lakh`), fully numeric, zero NaNs, and standardized numeric columns (mean ≈ 0, std ≈ 1).

